src/resolve/resolved-dns-transaction.c

```diff
@@ -2850,6 +2850,43 @@ static int dns_transaction_validate_dnskey_by_ds(DnsTransaction *t) {
         return 0;
 }
 
+typedef enum DelegationType {
+        DELEGATION_NONE,
+        DELEGATION_SECURE,
+        DELEGATION_INSECURE,
+} DelegationType;
+
+/* Evaluate the delegation type expressed in a transaction */
+static int dns_delegation_type(DnsTransaction *t) {
+        _cleanup_(dns_resource_key_unrefp) DnsResourceKey *ns = NULL;
+        int r;
+        DnssecNsecResult nsec_result;
+
+        assert(t);
+
+        /* Check that this is a answer to a query for a DS record */
+        if (dns_transaction_key(t)->type != DNS_TYPE_DS)
+                return 0;
+
+        r = dns_answer_match_key(t->answer, dns_transaction_key(t), /* ret_flags = */ NULL);
+        if (r < 0)
+                return r;
+
+        if (r > 0)
+                return DELEGATION_SECURE;
+
+        /* The answer didn't contain a DS record, check for proof of an NS record */
+        ns = dns_resource_key_new(dns_transaction_key(t)->class, DNS_TYPE_NS, dns_resource_key_name(dns_transaction_key(t)));
```

Comment:
Missing oom check. Also, please move the declaration of `ns` and `nsec_result` at the beginning of the function:
```c
_cleanup_(dns_resource_key_unrefp) DnsResourceKey *ns = 
                dns_resource_key_new(dns_transaction_key(t)->class, DNS_TYPE_NS, dns_resource_key_name(dns_transaction_key(t)));
if (!ns)
        return -ENOMEM;

DnssecNsecResult nsec_result;
r = dnssec_nsec_test(...);
```